### PR TITLE
Add support for geometrycollections to GeometryTreeReader and Writer

### PR DESCRIPTION
**Pull request by imotov**
Adds support geometry collections to GeometryTreeReader
and GeometryTreeWriter.

Relates #37206

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java [hunk 1]
```diff
@@ -34,42 +34,50 @@
  */
 public class GeometryTreeReader implements ShapeTreeReader {
 
-    private final int extentOffset = 8;
+    private static final int EXTENT_OFFSET = 8;
+    private final int extentPosition;
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: in the EdgeTree and such, this is called the `startPosition`. should we stay consistent here?

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java [hunk 2]
```diff
@@ -53,29 +55,56 @@ public GeometryTreeWriter(Geometry geometry, CoordinateEncoder coordinateEncoder
...
             extent = new Extent(builder.top, builder.bottom, builder.negLeft, builder.negRight, builder.posLeft, builder.posRight);
         }
         out.writeOptionalWriteable(extent);
         out.writeVInt(builder.shapeWriters.size());
-        for (ShapeTreeWriter writer : builder.shapeWriters) {
-            out.writeEnum(writer.getShapeType());
-            writer.writeTo(out);
+        if (multiShape) {
+            for (ShapeTreeWriter writer : builder.shapeWriters) {
+                try(BytesStreamOutput bytesStream = new BytesStreamOutput()) {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: Would it be just as readable and slightly more efficient to re-use one `BytesStreamOutput` and reset it?

Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: As we discussed I would like to figure out how to optimize this code here and in other places. I have a couple of ideas.

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java (hunk 1 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeReader.java]: Good catch! When I created this variable it was pointing to the start of extent.

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java (hunk 2 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: fine with punting this outside of this PR 👍

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java [hunk 3]
```diff
@@ -110,9 +139,7 @@ private void addWriter(ShapeTreeWriter writer) {
 
         @Override
         public Void visit(GeometryCollection<?> collection) {
-            for (Geometry geometry : collection) {
-                geometry.visit(this);
-            }
+            addWriter(new GeometryTreeWriter(collection, coordinateEncoder));
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: I looked at the GeoJSON spec, and it does not look like a geometrycollection can have a nested geometrycollection within it. I am not sure about WKT. do you know?

**Review by talevy — COMMENTED** [review]
Generally looks good to me, I just realized something about the nestedness of geometry-collection and wanted to be sure we are consistent here with regards to what we allow, and how we parse.

a side-note about the geometrycollection is that if someone indexes a geometry-collection with one item that is a geometrycollection, then we would be duplicating the centroids here as well.

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java (hunk 3 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: I think it is not as strong. According to RFC 7946

> To maximize interoperability, implementations SHOULD avoid nested GeometryCollections.

We support it elsewhere, so if we want to stop supporting it we should depreciate it first in other places.

**Comment by imotov** [comment]
Do you think we can do centroid optimization everywhere in another pass?

File: server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java (hunk 3 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java]: oh OK. good to know!

**Comment by talevy** [comment]
> Do you think we can do centroid optimization everywhere in another pass?

yes. that is reasonable!

**Review by talevy — APPROVED** [review]

**Pull request merged**